### Carga del archivo y ejecución de Preprocesamiento

In [13]:
import pandas as pd
import joblib
import sys, os

sys.path.append(os.path.abspath('../'))
from src.preprocessing import custom_feature_engineering

# RAW
df = pd.read_csv('../data/raw/03-bank_marketing.csv', sep=';')
df = custom_feature_engineering(df)

# X, y
X = df.drop("y", axis=1)
y = df["y"].apply(lambda x: 1 if x == "yes" else 0)

# split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)
preproc = joblib.load("../models/preprocessor.pkl")

### Entrenamiento de Modelos

In [14]:
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
import joblib
import os

os.makedirs("../models", exist_ok=True)

models = {
    "lr": LogisticRegression(max_iter=1000, random_state=42),
    "dt": DecisionTreeClassifier(random_state=42),
    "rf": RandomForestClassifier(random_state=42),
    "svm": SVC(),
    "knn": KNeighborsClassifier()
}

pipes = {}

for name, model in models.items():
    pipe = Pipeline([
        ("preproc", preproc),
        ("clf", model)
    ])
    
    pipe.fit(X_train, y_train)
    
    joblib.dump(pipe, f"../models/baseline_{name}.pkl")
    
    pipes[name] = pipe

print("Modelos entrenados y guardados correctamente")

Modelos entrenados y guardados correctamente


- #### Logistic Regression:
Es un modelo lineal que estima la probabilidad de pertenecer a una clase mediante una función logística. Funciona combinando las variables en una ecuación y aplicando una sigmoide para clasificar según un umbral. Se incluye como baseline por su simplicidad, rapidez e interpretabilidad.

- #### Decision Tree:
Es un modelo basado en reglas que divide los datos en forma de árbol mediante decisiones tipo if–else. Funciona separando iterativamente las observaciones según las variables que mejor discriminan las clases. Se incluye porque captura relaciones no lineales y es fácil de interpretar.

- #### Random Forest:
Es un ensamble de múltiples árboles de decisión que combina sus predicciones para mejorar el rendimiento. Funciona entrenando varios árboles con diferentes subconjuntos de datos y variables, reduciendo el sobreajuste. Se incluye por su alta capacidad de generalización y buen desempeño sin mucho ajuste.

- #### Support Vector Machine (SVM):
Es un modelo que busca separar las clases maximizando el margen entre ellas mediante un hiperplano. Puede usar funciones kernel para manejar relaciones no lineales. Se incluye por su robustez en espacios de alta dimensión y como comparación frente a otros enfoques.

- #### K-Nearest Neighbors (KNN):
Es un modelo basado en similitud que clasifica según la clase predominante de los vecinos más cercanos. Funciona midiendo distancias entre observaciones en el espacio de características. Se incluye por su simplicidad y como referencia frente a modelos más complejos.

### Evaluación

In [16]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

results = []

for name, pipe in pipes.items():
    y_pred = pipe.predict(X_test)
    
    results.append({
        "modelo": name,
        "accuracy": accuracy_score(y_test, y_pred),
        "precision": precision_score(y_test, y_pred),
        "recall": recall_score(y_test, y_pred),
        "f1": f1_score(y_test, y_pred)
    })

import pandas as pd
df_results = pd.DataFrame(results).sort_values(by="f1", ascending=False)
df_results

,modelo,accuracy,precision,recall,f1
2,rf,0.914764,0.663768,0.493534,0.566131
0,lr,0.910151,0.646875,0.446121,0.528061
1,dt,0.889145,0.507853,0.522629,0.515135
3,svm,0.911365,0.686792,0.392241,0.499314
4,knn,0.902380,0.606897,0.379310,0.466844


#### Comentario
  
Se entrenaron cinco modelos baseline utilizando un pipeline de preprocesamiento común para garantizar consistencia en la evaluación. El modelo Random Forest obtuvo el mejor desempeño general con un F1-score de 0.566, destacando en el equilibrio entre precisión y recall.

Dado que el problema presenta un fuerte desbalance de clases, el accuracy no es una métrica adecuada; por ello, se priorizó el uso del F1-score y el recall de la clase positiva, alineados con el objetivo del banco de identificar la mayor cantidad posible de clientes con alta probabilidad de aceptación del producto.

En este contexto, Random Forest permite mejorar la segmentación de clientes, enfocando los esfuerzos del call center en perfiles con mayor probabilidad de aceptación. Esto contribuye directamente a reducir llamadas innecesarias, aumentar la tasa de aceptación del depósito a plazo fijo y proteger la experiencia del cliente.

En consecuencia, se selecciona Random Forest como el modelo baseline más sólido para futuras etapas de optimización y despliegue.